### ISSUE TITLE
Can CaffeOnSpark used for deployment?

### ISSUE THREAD

**Issue by baoruxiao** [yahoo/CaffeOnSpark#248]
Hi, 

I have minor knowledge on spark and recently am exporing CaffeOnSpark framework.

I notice that caffeonspark read dataframe as datasource, and do some partitions on it and distribute partitions to executors. What if datasource contains single image?
I wonder if caffeonspark can read single image at a time and predict its label in a distributed manner? Namely, can I speed up predicting single image's label using caffeonspark clusters?

I have intention to use framework for DL model deployment, but this question lingered around and I'm reaching out for insights.

Many thanks!!

**Comment by junshi15** [comment]
If a dataset contains only one image, then only one partition will have an image, the rest are empty.

You can set batch size = 1 during inference (prediction) stage. This is actually recommended. If the batch size does not divide total dataset size, the reminder will be discarded as of current implementation. This is OK in training, but not desirable in inference. So at inference, you may want to set the batch size to 1, or make sure the batch size divides the total number of images.

**Comment by baoruxiao** [comment]
Thanks, this really answers my question.
However, based on your answer, if there is one partition, the inference performance will not horizontal scale with increasing of executors. Am I right?

**Comment by junshi15** [comment]
You want to do on-line (maybe even real-time) inference. CaffeOnSpark does not quite meet your requirement. 

A typical use case for CaffeOnSpark is as follows. You have a ton of images sitting on HDFS, you want to predict all of them. You launch CaffeOnSpark, say with N executors, the dataset gets cut into N partitions. Each executor work on one partition. As you can see, this is a batch operation. You don't change the number of executors during prediction. It's fixed for the entire process.

**Comment by baoruxiao** [comment]
@junshi15 This finally confirmed my speculation.  It seems no available clustering-solution to do on-line inference currently. Maybe virtualization is a way to go. 
Thanks! I will close this issue.

**Comment by junshi15** [comment]
You may want to look into tensorflow serving, if serving is your goal.

**Comment by baoruxiao** [comment]
@junshi15 Yes, serving/production is my goal. Seems really what I want, many thanks!!
